File: recursive-bubblesort/java/BubbleSort.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;

public class BubbleSort {

    public static void main(final String[] args){
        // parse the arguments
        final String filename = parseArgs(args);

        // extract file content from the filename
        final String[] input = readFile(filename);

        // sort the file content with a recursive bubble sort algorithm
        bubbleSort(input, input.length);

        // tally the sorted file content to replace duplicates with number of occurrences
        final Map<String, Integer> result = tallyWords(input);

        // print the sorted and tallied result
        System.out.println(result);
    }

    private static void bubbleSort(final String[] data, final int length){
        if(length == 1){
            return;
        }

        int count = 0;
        for(int i = 0; i < length - 1; i++){
            final String curVal = data[i];
            final String nxtVal = data[i + 1];

            if(Comparator.comparing(String::toString).compare(curVal, nxtVal) > 0){
                data[i] = nxtVal;
                data[i + 1] = curVal;
                count = count + 1;
            }
        }
        if(count == 0){
            return;
        }
        bubbleSort(data, length - 1);
    }

    private static String parseArgs(final String[] args){
        if(args.length != 1){
            throw new IllegalArgumentException("expected 'args.length' to be 1");
        }
        return args[0];
    }

    private static String[] readFile(final String filename){
        try {
            return Files.readAllLines(Paths.get(filename)).toArray(new String[0]);

        } catch (IOException e) {
            throw new IllegalArgumentException(e);
        }
    }

    private static Map<String, Integer> tallyWords(final String[] data){
        final Map<String, Integer> dataMap = new LinkedHashMap<>();
        for(String s : data){
            int curVal = dataMap.getOrDefault(s, 0);
            dataMap.put(s, curVal + 1);
        }
        return dataMap;
    }
}

File: recursive-bubblesort/cpp/bubblesort/BubbleSort.cpp
//
// Created by wheel on 7/31/2022.
//

#include <string>
#include <iostream>

using std::string;

#include "BubbleSort.h"

void BubbleSort::sort(vector<string>& data, int length) {
    if(length == 1){
        return;
    }
    auto count = 0;
    for(auto i = 0; i < length - 1; i++){
        const auto curVal = data[i];
        const auto nxtVal = data[i + 1];

        if(curVal.compare(nxtVal) > 0){
            data[i] = nxtVal;
            data[i + 1] = curVal;
            count++;
        }
    }
    if(count == 0){
        return;
    }
    sort(data, length - 1);
}

vector<std::pair<string, int>> BubbleSort::tally(vector<string> &data) {
    auto dataMapIndex = 0;
    std::pair<string, int> record;
    record.first = data[0];
    record.second = 1;

    vector<std::pair<string, int>> dataMap;
    dataMap.push_back(record);

    for (auto i = 1; i < data.size(); i++){
        std::pair<string, int>& prev = dataMap[dataMapIndex];
        if (prev.first == data[i]){
            prev.second++;
            continue;
        }
        record.first = data[i];
        record.second = 1;
        dataMap.push_back(record);
        dataMapIndex++;
    }
    return dataMap;
}
